### PR TITLE
make `txn FirstValidTime` actually the time of FirstValid-1

### PR DESCRIPTION
**Pull request by algobolson**
close hole where FirstValidTime of the block currently being assembled is hard to pin down

### CODE REVIEW

File: ledger/eval.go [hunk 1]
```diff
@@ -587,24 +587,23 @@ func (eval *BlockEvaluator) transaction(txn transactions.SignedTxn, ad transacti
...
-		Proto:               &eval.proto,
-		TxnGroup:            txgroup,
-		GroupIndex:          groupIndex,
-		FirstValidTimeStamp: uint64(hdr.TimeStamp),
+		Txn:        &txn,
+		Proto:      &eval.proto,
+		TxnGroup:   txgroup,
+		GroupIndex: groupIndex,
+	}
+	if hdr.TimeStamp < 0 {
```
Comment (tsachiherman) [inline ledger/eval.go]: if we're already picking up a date, why not to pick a more recent one ?

File: ledger/eval.go [hunk 2]
```diff
@@ -587,24 +587,23 @@ func (eval *BlockEvaluator) transaction(txn transactions.SignedTxn, ad transacti
 }
 
 func (eval *BlockEvaluator) checkLogicSig(txn transactions.SignedTxn, txgroup []transactions.SignedTxnWithAD, groupIndex int) (err error) {
-	firstValid := basics.Round(txn.Txn.FirstValid)
+	firstValid := basics.Round(txn.Txn.FirstValid - 1)
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply ledger/eval.go]: nit: change the variable name into "preFirstValid" to avoid confusion.

File: ledger/eval.go (hunk 1 above)
Comment (algobolson) [reply ledger/eval.go]: this is just to check that the cast from int64 to uint64 is clean

**Comment by algobolson** [comment]
this passes tests and I believe issues are addressed. re-check?

File: ledger/eval.go [hunk 3]
```diff
@@ -587,24 +587,24 @@ func (eval *BlockEvaluator) transaction(txn transactions.SignedTxn, ad transacti
...
-		if err != nil {
-			return fmt.Errorf("could not fetch BlockHdr for FirstValid=%d (current=%d): %s", txn.Txn.FirstValid, eval.block.BlockHeader.Round, err)
-		}
+	if txn.Txn.FirstValid == 0 {
+		return errors.New("LogicSig does not work with FirstValid==0")
+	}
+	// TODO: move this into some lazy evaluator for the few scripts that actually use `txn FirstValidTime` ?
+	hdr, err := eval.l.BlockHdr(basics.Round(txn.Txn.FirstValid - 1))
+	if err != nil {
+		return fmt.Errorf("could not fetch BlockHdr for FirstValid-1=%d (current=%d): %s", txn.Txn.FirstValid-1, eval.block.BlockHeader.Round, err)
```
Comment (tsachiherman) [inline ledger/eval.go]: nit: last formatting object should be `%v` rather than `%s`

**Pull request merged**